x-pack/plugins/observability/server/services/slo/resource_installer.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  ClusterPutComponentTemplateRequest,
+  IndicesPutIndexTemplateRequest,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { ElasticsearchClient, Logger } from '@kbn/core/server';
+
+import {
+  SLO_COMPONENT_TEMPLATE_MAPPINGS_NAME,
+  SLO_COMPONENT_TEMPLATE_SETTINGS_NAME,
+  SLO_INDEX_TEMPLATE_NAME,
+} from '../../assets/constants';
+import { sloMappingsTemplate } from '../../assets/component_templates/slo_mappings_template';
+import { sloSettingsTemplate } from '../../assets/component_templates/slo_settings_template';
+import { getSloIndexTemplate } from '../../assets/index_templates/slo_index_templates';
+
+export class ResourceInstaller {
+  constructor(private esClient: ElasticsearchClient, private logger: Logger) {}
+
+  public async ensureCommonResourcesInstalled(): Promise<void> {
+    const existingIndexTemplate = await this.checkIndexTemplateExists();
+
+    if (existingIndexTemplate) {
+      this.logger.debug(`Skipping installation of resources shared between all SLO indices`);
+      return;
+    }
+
+    try {
+      await Promise.all([
+        this.createOrUpdateComponentTemplate({
+          name: SLO_COMPONENT_TEMPLATE_MAPPINGS_NAME,
+          ...sloMappingsTemplate,
+        }),
+
+        this.createOrUpdateComponentTemplate({
+          name: SLO_COMPONENT_TEMPLATE_SETTINGS_NAME,
+          ...sloSettingsTemplate,
+        }),
+      ]);
+
+      await this.createOrUpdateIndexTemplate({
```

Comment:
I wonder on an enterprise level if the mapping and the template will be ready and propagated when running `createOrUpdateIndexTemplate` just right after creating them? is that a valid concern @simianhacker ?

Do we need a simple retry mechanism inside `createOrUpdateIndexTemplate `?

Reply:
We could add some retry here or let the client retries the API call if it fails the first time as well. This resource installation should only happen with the first SLO creation.

I'm fine with adding a retry here if we feel the need to. Let me know

Reply:
In my testing, they are immediately available.